File: typescript/packages/core/src/core/github/read.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { GitHubAccessor } from '../../accessor/github.ts'
import type { IndexCacheStore } from '../../cache/index/store.ts'
import type { PathSpec } from '../../types.ts'
import { fetchBlob } from './_client.ts'
import { stripSlash } from '../../util/slash.ts'

function enoent(path: string): Error {
  const e = new Error(`ENOENT: ${path}`) as Error & { code: string }
  e.code = 'ENOENT'
  return e
}

function eisdir(path: string): Error {
  const e = new Error(`EISDIR: ${path}`) as Error & { code: string }
  e.code = 'EISDIR'
  return e
}

function stripPrefix(path: PathSpec): string {
  const prefix = path.prefix
  let p = path.original
  if (prefix !== '' && p.startsWith(prefix)) {
    p = p.slice(prefix.length) || '/'
  }
  return p
}

function indexKey(p: string): string {
  const trimmed = stripSlash(p)
  return trimmed === '' ? '/' : `/${trimmed}`
}

export async function read(
  accessor: GitHubAccessor,
  path: PathSpec,
  index?: IndexCacheStore,
): Promise<Uint8Array> {
  const p = stripPrefix(path)
  if (index === undefined) throw enoent(p)
  const result = await index.get(indexKey(p))
  if (result.entry === undefined || result.entry === null) throw enoent(p)
  if (result.entry.resourceType === 'folder') throw eisdir(p)
  return fetchBlob(accessor.transport, accessor.owner, accessor.repo, result.entry.id)
}

export async function readBytes(accessor: GitHubAccessor, sha: string): Promise<Uint8Array> {
  return fetchBlob(accessor.transport, accessor.owner, accessor.repo, sha)
}

export async function* stream(
  accessor: GitHubAccessor,
  path: PathSpec,
  index?: IndexCacheStore,
): AsyncIterable<Uint8Array> {
  const data = await read(accessor, path, index)
  yield data
}

File: python/mirage/core/github/read.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import base64

from mirage.accessor.github import GitHubAccessor
from mirage.cache.index import IndexCacheStore, LookupStatus
from mirage.core.github._client import github_get
from mirage.core.github.config import GitHubConfig
from mirage.types import PathSpec
from mirage.utils.errors import enoent


async def read_bytes(config: GitHubConfig, owner: str, repo: str,
                     sha: str) -> bytes:
    data = await github_get(
        config.token,
        "/repos/{owner}/{repo}/git/blobs/{sha}",
        owner=owner,
        repo=repo,
        sha=sha,
    )
    return base64.b64decode(data["content"])


async def read(
    accessor: GitHubAccessor,
    path: PathSpec,
    index: IndexCacheStore = None,
) -> bytes:
    if isinstance(path, str):
        path = PathSpec(original=path, directory=path)
    virtual = path.original
    if isinstance(path, PathSpec):
        prefix = path.prefix
        path = path.original

    if prefix and path.startswith(prefix):
        rest = path[len(prefix):]
        if prefix.endswith("/") or rest == "" or rest.startswith("/"):
            path = rest or "/"
    key = "/" + path.strip("/")
    if index is None:
        raise enoent(virtual)
    result = await index.get(key)
    if result.status == LookupStatus.NOT_FOUND or result.entry is None:
        raise enoent(virtual)
    if result.entry.resource_type == "folder":
        raise IsADirectoryError(virtual)
    return await read_bytes(accessor.config, accessor.owner, accessor.repo,
                            result.entry.id)
